test/test-udp-reuseport.c

```diff
@@ -0,0 +1,296 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#if !defined(__linux__) && !defined(__FreeBSD__) && \
+    !defined(__DragonFly__) && !defined(__sun) && !defined(_AIX73)
+
+TEST_IMPL(udp_reuseport) {
+  struct sockaddr_in addr1, addr2, addr3;
+  uv_loop_t* loop;
+  uv_udp_t handle1, handle2, handle3;
+  int r;
+
+  ASSERT_OK(uv_ip4_addr("127.0.0.1", TEST_PORT, &addr1));
+  ASSERT_OK(uv_ip4_addr("127.0.0.1", TEST_PORT_2, &addr2));
+  ASSERT_OK(uv_ip4_addr("127.0.0.1", TEST_PORT_3, &addr3));
+
+  loop = uv_default_loop();
+  ASSERT_NOT_NULL(loop);
+
+  r = uv_udp_init(loop, &handle1);
+  ASSERT_OK(r);
+
+  r = uv_udp_bind(&handle1, (const struct sockaddr*) &addr1, UV_UDP_REUSEADDR);
+  ASSERT_OK(r);
+
+  r = uv_udp_init(loop, &handle2);
+  ASSERT_OK(r);
+
+  r = uv_udp_bind(&handle2, (const struct sockaddr*) &addr2, UV_UDP_REUSEPORT);
+  ASSERT_EQ(r, UV_ENOTSUP);
+
+  r = uv_udp_init(loop, &handle3);
+  ASSERT_OK(r);
+
+  /* For platforms where SO_REUSEPORTs don't have the capability of
+   * load balancing, specifying both UV_UDP_REUSEADDR and UV_UDP_REUSEPORT
+   * in flags will fail, returning an UV_ENOTSUP error. */
+  r = uv_udp_bind(&handle3, (const struct sockaddr*) &addr3,
+                  UV_UDP_REUSEADDR | UV_UDP_REUSEPORT);
+  ASSERT_EQ(r, UV_ENOTSUP);
+
+  MAKE_VALGRIND_HAPPY(loop);
+
+  return 0;
+}
+
+#else
+
+#define NUM_RECEIVING_THREADS 2
+#define MAX_UDP_DATAGRAMS 10
+
+static uv_udp_t udp_send_handles[MAX_UDP_DATAGRAMS];
+static uv_udp_send_t udp_send_requests[MAX_UDP_DATAGRAMS];
+
+static uv_sem_t semaphore;
+
+static uv_mutex_t mutex;
+static unsigned int received;
+
+static unsigned int thread_loop1_recv;
+static unsigned int thread_loop2_recv;
+static unsigned int sent;
+
+static uv_loop_t* main_loop;
+static uv_loop_t* thread_loop1;
+static uv_loop_t* thread_loop2;
```

Comment:
```suggestion
static uv_loop_t thread_loop1;
static uv_loop_t thread_loop2;
```
Nit: don't malloc the loops would make the code cleaner